src/transformStateWithClones.js

```diff
@@ -7,7 +7,47 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let curentState = { ...state };
+
+  for (const action of actions) {
+    const { type = null, extraData = {}, keysToRemove = [] } = action;
+
+    switch (type) {
+      case 'addProperties':
+        addProperties(curentState, extraData);
+        break;
+      case 'removeProperties':
+        removeProperties(curentState, keysToRemove);
+        break;
+      case 'clear':
+        clear(curentState);
+        break;
+    }
+
+    stateHistory.push(curentState);
+
+    curentState = { ...curentState };
+  }
+
+  return stateHistory;
+}
+
+function addProperties(state, extraData) {
+  Object.assign(state, extraData);
+}
+
+function removeProperties(state, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete state[key];
+  }
```

Comment:
The 'removeProperties' function is mutating the passed 'state' object. Instead, it should return a new object with the specified keys removed.